src/core/exception.d

```diff
@@ -72,6 +77,143 @@ unittest
     }
 }
 
+/**
+ * Thrown when an out of bounds array index is accessed.
+ */
+class ArrayIndexError : RangeError
+{
+    /// Index into array
+    const size_t index;
+    /// Length of indexed array
+    const size_t length;
+
+    // Buffer to avoid GC allocations
+    private immutable char[100] msgBuf = '\0';
+
+    this(size_t index, size_t length, string file = __FILE__,
+         size_t line = __LINE__, Throwable next = null) @nogc nothrow pure @safe
+    {
+        this.index  = index;
+        this.length = length;
+
+        // Constructing the message is a bit clumsy:
+        // It's essentially `printf("index [%zu] exceeds array length [%zu]", index, length)`,
+        // but even `snprintf` isn't `pure`.
+        // Also string concatenation isn't `@nogc`, and casting to/from immutable isn't `@safe`
+        import core.internal.string : unsignedToTempString;
+        size_t p = 0;
+        char[msgBuf.length] buf = void;
+        char[20] tmpBuf = void;
+        void put(const(char)[] s)
+        {
+            assert(p + s.length <= buf.length); // don't throw ArraySliceError inside ArrayIndexError ctor
+            buf[p..p+s.length] = s[];
+            p += s.length;
+        }
```

Comment:
Maybe this should be supported by `core.internal.string.TempStringNoAlloc` to avoid the duplication here (and probably for `-checkaction=context` in the future)

Reply:
You mean `put` should be a free function in `core.internal.string`?

Reply:
Or a member function. Wtatever works best to reduce the duplication.

Reply:
Shaved off 5 lines now